### PR TITLE
[Fixtures] Address factory

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Branch?         | foundry-fixtures
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        return [
+            'first_name' => self::faker()->firstName(),
+            'last_name' => self::faker()->lastName(),
+            'phone_number' => self::faker()->boolean() ? self::faker()->phoneNumber() : null,
+            'company' => self::faker()->boolean() ? self::faker()->company() : null,
+            'street' => self::faker()->streetAddress(),
+            'city' => self::faker()->city(),
+            'postcode' => self::faker()->postcode(),
+            'country_code' => $this->countryFactory::randomOrCreate()->getCode(),
```
Comment (Roshyo) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php:121]: Why are we calling a static method from an object ?
Wouldn't it be easier to just not inject the dependency and call the Static from class name ?

Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php:121]: We do not use parameters for the class names of the factories. When using your own factory service, you just use a service alias.

```yaml
services:
   Sylius\Bundle\CoreBundle\DataFixtures\Factory\CountryFactoryInterface: App\DataFixtures\Factor\CountryFactory
```

For what you suggest, you have to change a parameter for the factory class like this:
```yaml
parameters:
    sylius.data_fixtures.factory.country.class: App\DataFixtures\Factor\CountryFactory
```

But I think it's the legacy way and not Symfony-like. I didn't choose this type of DI configuration.

**Review by Roshyo — APPROVED** [review]

**Review by vvasiloi — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            'province_name' => null,
+            'province_code' => null,
+            'customer' => null,
```
[the remark this answers is not in the document]
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php]: done

**Review by Roshyo — APPROVED** [review]

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    public function withProvinceCode(string $provinceCode): self
+    {
+        return $this->addState(['province_code' => $provinceCode]);
+    }
+
+    public function withCustomer(Proxy|CustomerInterface|string $customer): self
+    {
+        return $this->addState(function () use ($customer): array {
+            if (is_string($customer)) {
+                $customer = $this->shopUserFactory::randomOrCreate(['email' => $customer])->getCustomer();
```
Comment (vvasiloi) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/AddressFactory.php:104]:
```suggestion
                $customer = $this->shopUserFactory::findOrCreate(['email' => $customer])->getCustomer();
```

**Review by vvasiloi — CHANGES_REQUESTED** [review]

**Pull request merged**